docs: how we handle rare build system (#14831)
conan-io/conan-center-index: docs/faqs.md

We generally consider tools like CMake as a standard tool to have installed in your system. Having the `cmake` package as a build require in **all** the recipes that use it will be an overkill, as every build requirement is installed like a requirement and takes time to download. However, `cmake` could still be useful to use in your profile:

```
[tool_requires]
cmake/3.17.2
```

Other packages using more unusual build tools, like `OpenSSL` using `strawberryperl`, will have the build require in the recipe as it is likely that the user that want to build it from sources will not have it installed in their system.

## How are rare build systems without generators packaged?

The C++ ecosystem has a lot of rare, unique and obscure build systems. Some of these are available in ConanCenter but they do not have built-in generators from the main Conan client.

> For these cases, contributors are asked to help reviewers as much as possible as it's likely we will not have expertise.

Reviewer: I think something more actionable is needed, like "the recipe must explicitly encode the specifics of the build system, mapping the ``settings``, ``options`` for the binary configuration, and also mapping ``self.dependencies`` so the build system can locate the dependencies libraries``, or something in that line? Maybe some link to some example recipe?

Reviewer 2: yes, I suppose we should illustrate with examples (links to existing recipes are okay) on how exactly do we:
- encode the specifics of the build system (I am personally not sure what does it mean, what specifics and how to encode?)
- mapping the settings, options for the binary configuration
- mapping self.dependencies so the build system can locate the dependencies libraries

in general, we should try our best to avoid vague/abstracts statements on **what** to do, instead explaining **how** to do things specifically. we cannot expect all our readers, especially newcomers, to know about all these things.

Author: Are there up to date docs for writing generators? https://docs.conan.io/en/latest/reference/generators/custom.html?highlight=generators https://docs.conan.io/en/latest/howtos/custom_generators.html#dyn-generators seem like the old stuff

https://docs.conan.io/en/2.0/search.html?q=generators&check_keywords=yes&area=default Nothing in the 2.0 docs

Author: ```suggestion
The recipe is expected to encode the specifics of the build system, mapping the `settings`, `options` for the binary configuration, and also mapping `self.dependencies` so the build system can locate the dependencies libraries as required.
For these cases, contributors are asked to help reviewers as much as possible as it's likely we will not have expertise.

> TODO: Add a link to docs.conan.io which explains how to write a custom generator in the 2.0 sense
```

> For quality assurance the build service is expected to be green and the hooks will ensure the package contents match whats expected given the options. This means they are more likely to have

Reviewer 2: who are `they` here? hooks? package contents? options? build services?

> inconsistency then other projects but make for excellent contributons.

Reviewer 2: you need to rephrase that... I fail to parse that statement. did you mean something like `inconsistency with other projects but make excellent contributions.`?

Author: ```suggestion
inconsistency with other recipes but make for excellent contributions.
```

> For quality assurance the build service is expected to be green and the [hooks](https://github.com/conan-io/hooks) will ensure the package contents match what is expected given the options. This means they are more likely to have

Reviewer 2: somehow my question got erased from the diff...
who are `they` here? hooks? package contents? options? build services?

> For quality assurance the build service is expected to be green and the [hooks](https://github.com/conan-io/hooks) will ensure the package contents match what is expected given the options. This means they are more likely to have

Author: ```suggestion
For quality assurance the build service is expected to be green and the [hooks](https://github.com/conan-io/hooks) will ensure the package contents match what is expected given the options. These recipes are more likely to have
```

We generally consider tools like CMake as a standard tool to have installed in your system. Having the `cmake` package as a build require in **all** the recipes that use it will be an overkill, as every build requirement is installed like a requirement and takes time to download. However, `cmake` could still be useful to use in your profile:

```
[tool_requires]
cmake/3.17.2
```

> Other packages using more unusual build tools, like [`OpenSSL`](https://github.com/conan-io/conan-center-index/tree/master/recipes/openssl) using [`strawberryperl`](https://github.com/conan-io/conan-center-index/tree/master/recipes/strawberryperl), will have the build require in the recipe as it is likely that the user that want to build it from sources will not have it installed in their system.

Author: ```suggestion
Other packages using more unusual build tools should refer to the [Dependencies - Adding Build Requirements](adding_packages/dependencies.md#build-requirements) section for more information.
```
